File: day2/d2p2.py
def is_safe(report):
    increasing = all(report[i] < report[i + 1] and 1 <= report[i + 1] - report[i] <= 3 for i in range(len(report) - 1))
    decreasing = all(report[i] > report[i + 1] and 1 <= report[i] - report[i + 1] <= 3 for i in range(len(report) - 1))
    return increasing or decreasing

def can_be_safe_by_removing_one(report):
    for i in range(len(report)):
        new_report = report[:i] + report[i+1:]
        if is_safe(new_report):
            return True
    return False

# Step 1: Read the data from d2input.txt
with open('d2input.txt', 'r') as file:
    lines = file.readlines()

# Step 2: Split each line into a list of numbers
reports = [list(map(int, line.split())) for line in lines]

# Step 3-6: Check each report and count the number of safe reports
safe_count = 0
for report in reports:
    if is_safe(report) or can_be_safe_by_removing_one(report):
        safe_count += 1

print("Number of safe reports:", safe_count)

File: day2/d2p2.js
const fs = require('fs');

// Function to check if a report is safe
function isSafe(report) {
    let increasing = true;
    let decreasing = true;

    for (let i = 0; i < report.length - 1; i++) {
        let diff = report[i + 1] - report[i];
        if (diff < 1 || diff > 3) {
            increasing = false;
        }
        if (diff > -1 || diff < -3) {
            decreasing = false;
        }
    }

    return increasing || decreasing;
}

// Function to check if a report can be made safe by removing one level
function canBeSafeByRemovingOne(report) {
    for (let i = 0; i < report.length; i++) {
        let newReport = report.slice(0, i).concat(report.slice(i + 1));
        if (isSafe(newReport)) {
            return true;
        }
    }
    return false;
}

// Step 1: Read the data from d2input.txt
const data = fs.readFileSync('d2input.txt', 'utf8');
const lines = data.trim().split('\n');

// Step 2: Split each line into a list of numbers
let reports = lines.map(line => line.trim().split(/\s+/).map(Number));

// Step 3-6: Check each report and count the number of safe reports
let safeCount = 0;
for (let report of reports) {
    if (isSafe(report) || canBeSafeByRemovingOne(report)) {
        safeCount++;
    }
}

console.log("Number of safe reports:", safeCount);